### ISSUE TITLE
.NET6 build nondeterministic and fails every 5/6 attempts

### ISSUE THREAD

**Issue by DmitryJustDmitry** [dotnet/source-build#2750]
Hi, everyone.

### Bug Description

We've being bumped in a very strange situation since the beginning of 2022 - those days we tried to build .NET6 for the first time (now it's ~70th attempt). We build it in docker environment - host system is Ubuntu 20.04, AMD 32cores/256RAM/1TB and Intel 16cores/64RAM/256GB  (different hosts). **The problem is a build nondeterminism - only 1/6 of build attempts succeeds.**

We've prepared a Dockerfile based on Fedora34 (because Fedora33 recommended in [README.md](https://github.com/dotnet/source-build#readme) told us that GLIBC version is not compatible) and recommended [prerequisites](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/tree/main/src/fedora/34/amd64) - the Dockerfile is attached below. It downloads all the repos, makes tarball and preps it for `build.sh`. Then we try to execute `build.sh` and sometimes (1/6 of attempts) the build succeeds:

```bash
...
"/sources/build.proj" (default target) (1) ->
"/sources/repos/known-good.proj" (Build target) (3) ->
  /sources/repos/Directory.Build.targets(364,5): warning : runtime.linux-x64.Microsoft.NETCore.ILAsm/6.0.0 : Nupkg found at '/sources/packages/previously-source-built/runtime.linux-x64.Microsoft.NETCore.ILAsm.6.0.0.nupkg' [/sources/repos/known-good.proj]


"/sources/build.proj" (default target) (1) ->
"/sources/repos/known-good.proj" (Build target) (3) ->
  /sources/repos/Directory.Build.targets(364,5): warning : runtime.linux-x64.Microsoft.NETCore.ILDAsm/6.0.0 : Nupkg found at '/sources/packages/previously-source-built/runtime.linux-x64.Microsoft.NETCore.ILDAsm.6.0.0.nupkg' [/sources/repos/known-good.proj]


"/sources/build.proj" (default target) (1) ->
"/sources/repos/known-good.proj" (Build target) (3) ->
  /sources/repos/Directory.Build.targets(364,5): warning : System.Reflection.MetadataLoadContext/6.0.0 : Nupkg found at '/sources/packages/previously-source-built/System.Reflection.MetadataLoadContext.6.0.0.nupkg' [/sources/repos/known-good.proj]


"/sources/build.proj" (default target) (1) ->
"/sources/repos/known-good.proj" (WritePrebuiltUsageData;ReportPrebuiltUsage target) (3:25) ->
(ReportPrebuiltUsage target) ->
  /sources/repos/Directory.Build.targets(834,5): warning : Prebuilt usages are different from the baseline. If detected changes are acceptable, update baseline with: [/sources/repos/known-good.proj]
/sources/repos/Directory.Build.targets(834,5): warning : cp '/sources/artifacts/prebuilt-report/generated-new-baseline.xml' '/sources/tools-local/prebuilt-baseline.xml' [/sources/repos/known-good.proj]

    275 Warning(s)
    0 Error(s)

Time Elapsed 00:46:10.74
```

But most of the time it fails. Everytime in the `runtime`  package build phase. It looks like:

```bash
"/sources/build.proj" (default target) (1) ->
"/sources/repos/known-good.proj" (Build target) (3) ->
"/sources/repos/msbuild.proj" (Build target) (17:9) ->
"/sources/repos/runtime.proj" (Build target) (23:9) ->
(RepoBuild target) ->
  /sources/repos/Directory.Build.targets(418,5): error MSB3073: The command "/sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/build.sh  --ci --configuration Release --restore --build --pack --publish -bl /p:ArcadeBuildFromSource=true /p:CopyWipIntoInnerSourceBuildRepo=true /p:DotNetBuildOffline=true /p:CopySrcInsteadOfClone=true /p:DotNetPackageVersionPropsPath="/sources/artifacts/obj/x64/Release/PackageVersions.props" /p:AdditionalSourceBuiltNupkgCacheDir="/sources/artifacts/obj/x64/Release/blob-feed/packages/" /p:ReferencePackageNupkgCacheDir="/sources/packages/reference/packages/" /p:PreviouslySourceBuiltNupkgCacheDir="/sources/packages/previously-source-built/" /p:TargetRid=fedora.34-x64 /p:SourceBuildNonPortable=true  /p:DotNetPackageVersionPropsPath=/sources/artifacts/obj/x64/Release/PackageVersions.props /p:DotNetRestoreSourcePropsPath=/sources/artifacts/obj/x64/Release/RestoreSources.props /p:DotNetBuildOffline=true >> /sources/artifacts/logs/runtime.log 2>&1" exited with code 1. [/sources/repos/runtime.proj]

    47 Warning(s)
    1 Error(s)

Time Elapsed 00:26:41.98
```

The first errors in builder log looks like:

```bash
Successfully created package '/sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/artifacts/packages/Release/Shipping/Microsoft.Extensions.Options.6.0.0.symbols.nupkg'.
      System.Data.Odbc -> /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/artifacts/bin/System.Data.Odbc/net6.0-Release/System.Data.Odbc.dll
      Microsoft.Extensions.Logging.EventSource -> /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/artifacts/bin/Microsoft.Extensions.Logging.EventSource/net461-Release/Microsoft.Extensions.Logging.EventSource.dll
      System.Data.OleDb -> /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/artifacts/bin/System.Data.OleDb/net6.0-Release/System.Data.OleDb.dll
      System.Security.Cryptography.Pkcs -> /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/artifacts/bin/System.Security.Cryptography.Pkcs/netstandard2.0-windows-Release/System.Security.Cryptography.Pkcs.dll
    /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/package-cache/microsoft.dotnet.compatibility/1.0.0-rtm.21521.8/build/Microsoft.NET.Compatibility.Common.targets(32,5): error MSB4018: The "Microsoft.DotNet.Compatibility.ValidatePackage" task failed unexpectedly. [/sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/src/libraries/Microsoft.Extensions.Options/src/Microsoft.Extensions.Options.csproj]
    /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/package-cache/microsoft.dotnet.compatibility/1.0.0-rtm.21521.8/build/Microsoft.NET.Compatibility.Common.targets(32,5): error MSB4018: System.MissingMethodException: Method not found: 'Int32 Microsoft.CodeAnalysis.ISymbol.get_MetadataToken()'. [/sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/src/libraries/Microsoft.Extensions.Options/src/Microsoft.Extensions.Options.csproj]
    /sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/package-cache/microsoft.dotnet.compatibility/1.0.0-rtm.21521.8/build/Microsoft.NET.Compatibility.Common.targets(32,5): error MSB4018:    at Microsoft.CodeAnalysis.CSharp.Symbols.AssemblySymbol.CreateISymbol() [/sources/src/runtime.4822e3c3aa77eb82b2fb33c9321f923cf11ddde6/artifacts/source-build/self/src/src/libraries/Microsoft.Extensions.Options/src/Microsoft.Extensions.Options.csproj]
```

As for me it's very strange, because we think that build process is based on a complete tarball, and no side effects should present on constant sources in a constant container. We tried to recreate docker images (3 times) but this effect repeats for a month - after 70 attempts we can`t create deterministic build process, unfortunately. It sounds very strange, but it's a fact we really have.

Please check the Dockerfile, maybe it has an error we've not noticed? Try to build and run it for 7-5 times - some of your attempts should end with a build error at `portable` build stage. If someone could share a docker-image (in private or as a response) that guarantee deterministic build, it would be perfect! Thanks!  

### Steps to Reproduce

```bash
docker build -f Dockerfile . -t fed34net6:ready_to_build
docker run -it --name=net6build0 fed34net6:ready_to_build /bin/bash
(in container) cd /sources && ./build.sh
```

The Dockerfile is [Dockerfile.zip](https://github.com/dotnet/source-build/files/8005121/Dockerfile.zip):

```dockerfile
FROM registry.fedoraproject.org/fedora:34

# Install the base toolchain we need to build anything (clang, cmake, make and the like)
# this does not include libraries that we need to compile different projects, we'd like
# them in a different layer.
RUN dnf --setopt=install_weak_deps=False install -y \
        clang \
        cmake \
        dnf-plugins-core \
        findutils \
        gdb \
        glibc-langpack-en \
        lldb-devel \
        llvm-devel \
        make \
        python \
        which \
    && dnf clean all

# Add MS package repo
COPY  microsoft.asc /tmp
RUN dnf config-manager --add-repo https://packages.microsoft.com/fedora/34/prod/ && \
    rpm --import /tmp/microsoft.asc && \
    rm /tmp/microsoft.asc

# Install tools used by build automation.
RUN dnf --setopt=install_weak_deps=False install -y \
        git \
        tar \
        procps \
        zip \
    && dnf clean all

# Dependencies of CoreCLR, Mono and CoreFX.
RUN dnf --setopt=install_weak_deps=False install -y \
        autoconf \
        automake \
        glibc-locale-source \
        iputils \
        krb5-devel \
        libcurl-devel \
        libgdiplus \
        libicu-devel \
        libomp-devel \
        libtool \
        libunwind-devel \
        libuuid-devel \
        lttng-ust-devel \
        openssl-devel \
        uuid-devel \
        zlib-devel \
    && dnf clean all

RUN cd / && \
git clone https://github.com/dotnet/installer && \
cd /installer && \
git switch -c v6.0.100 v6.0.100

RUN  cd /installer && \
./build.sh /p:ArcadeBuildTarball=true /p:TarballDir=/sources

RUN cd /sources && \
./prep.sh

#RUN ./build.sh
```

**Comment by omajid** [comment]
Hello!

The error you are running into sounds a lot like https://github.com/dotnet/source-build/issues/2501. https://github.com/dotnet/runtime/issues/59908 is also the same issue and it calls out how this error involves a race condition, which matches up with the non-determinism you are seeing.

For .NET 6.0, a "fix" was merged (https://github.com/dotnet/runtime/pull/60881) which just disables this unnecessary package validation check. The fix missed 6.0.100, but should have been included in 6.0.101.

Since you are using the `v6.0.100` tag, can you apply a manual hack? After creating `/sources` (before or after `prep.sh` shouldn't matter), can you add this step?

    sed -i -E 's|( /p:BuildDebPackage=false)|\1 /p:EnablePackageValidation=false|' src/runtime.*/eng/SourceBuild.props 

That should add a `/p:EnablePackageValidation=false` flag to the `SourceBuild.props` file under the runtime directory. That will turn off the unnecessary package validation and stop the build from failing. This is what we ended up doing for the Fedora .NET 6.0 packages and it has been working great so far.

---

Separately, please consider using `./prep.sh --bootstrap` instead of a plain `./prep.sh`. That gets you binaries that should work in *most* Linux environments (older and newer releases of Fedora too).

**Comment by DmitryJustDmitry** [comment]
Hi!

> The error you are running into sounds a lot like #2501. [dotnet/runtime#59908](https://github.com/dotnet/runtime/issues/59908) is also the same issue and it calls out how this error involves a race condition, which matches up with the non-determinism you are seeing.
> 
> For .NET 6.0, a "fix" was merged ([dotnet/runtime#60881](https://github.com/dotnet/runtime/pull/60881)) which just disables this unnecessary package validation check. The fix missed 6.0.100, but should have been included in 6.0.101.

Thanks a lot! I read that ones but didn't realize that it's a case of mine.
 
> Since you are using the `v6.0.100` tag, can you apply a manual hack? After creating `/sources` (before or after `prep.sh` shouldn't matter), can you add this step?
> 
> ```
> sed -i -E 's|( /p:BuildDebPackage=false)|\1 /p:EnablePackageValidation=false|' src/runtime.*/eng/SourceBuild.props 
> ```
> 
> That should add a `/p:EnablePackageValidation=false` flag to the `SourceBuild.props` file under the runtime directory. That will turn off the unnecessary package validation and stop the build from failing. This is what we ended up doing for the Fedora .NET 6.0 packages and it has been working great so far.

**Done and works well!** An extra bonus - got some information about new validation mechanism (thanks to Anirudh [article](https://devblogs.microsoft.com/dotnet/package-validation/)). 
 
> Separately, please consider using `./prep.sh --bootstrap` instead of a plain `./prep.sh`. That gets you binaries that should work in _most_ Linux environments (older and newer releases of Fedora too).

Thanks for your advice - done it, works well. Hope that in 6.0.1-3 (according to the Issue dialogues) all of this tricks will be implemented into the main branch.
